docs/src/internals.md

```diff
@@ -0,0 +1,256 @@
+# Internals and extending the formula DSL
+
+This section is intended to help **package developers** understand the internals
+of how a `@formula` becomes a numerical matrix, in order to use, manipulate, and
+even extend the DSL.  The Julia `@formula` is designed to be as extensible as
+possible through the normal Julian mechanisms of multiple dispatch.
+
+## The lifecycle of a `@formula`
+
+A formula goes through a number of stages, starting as an
+expression that's passed to the `@formula` macro and ending up generating a
+numeric matrix when ultimately combined with a tabular data source:
+
+1. "Macro time" when only the surface syntax is available
+2. "Schema time" incorporates information about **data invariants** (types of each
+   variable, levels of categorical variables, summary statistics for continuous
+   variables) and the **model type**.
+3. "Data time" when the actual data values themselves are available.
+
+For in-memory (columnar) tables, there is not much difference between "data
+time" and "schema time" in practice, but in principle it's important to
+distinguish between these when dealing with truly streaming data, or large data
+stores where calculating invariants of the data may be expensive.
+
+### Macro time
+
+The `@formula` macro does syntactic transformations of the formula expression.
+At this point, _only_ the expression itself is available, and there's no way to
+know whether a term corresponds to a continuous or categorical variable.
+
+For standard formulae, this amounts to applying the syntactic rules for the DSL
+operators (expanding `*` and applying the distributive and associative rules),
+and wrapping each symbol in a `Term` constructor:
+
+```julia-repl
+julia> @macroexpand @formula(y ~ 1 + a*b)
+:(Term(:y) ~ ConstantTerm(1) + Term(:a) + Term(:b) + Term(:a) & Term(:b))
+```
+
+Calling this stage "macro time" is a bit of a misnormer because much of the
```

Comment:
Maybe "syntax time" then?

Reply:
Yes I like that better...and it also jives better with the possibilty of getting a generated string too...